### PR TITLE
fix(compress): `toFixed`, `toPrecision`, `toExponential` and `toString` of Number

### LINKED ISSUE #6957: 'compress' option breaks number literals ending with '0' when .toFixed() is called
### Describe the bug

Calling .toFixed() on a numeric literal value with a number > 0 produces an optimized literal that is incorrect. It appears to remove trailing zeroes in the output such that if the decimal separator came after the first digit, the output would have the correct number of digits after the separator.

Using swc playground, toggling the 'Compress' option is the only flag I can find which is relevant to the output

### Input code

```typescript
//                                actual | expected
alert(1..toFixed(1))        //     '1.0'   '1.0'
alert(0..toFixed(0))        //       '0'   '0'
alert(0..toFixed(1))        //       '0'   '0.0'
alert(0..toFixed(2))        //       '0'   '0.00'
alert(0..toFixed(3))        //       '0'   '0.000'
alert(10..toFixed(1))       //      '10'   '10.0'
alert(20..toFixed(2))       //    '20.0'   '20.00' 
alert(30..toFixed(3))       //   '30.00'   '30.000' 
alert(100..toFixed(1))      //     '100'   '100.0'
alert(100..toFixed(2))      //     '100'   '100.00'
alert(100..toFixed(3))      //   '100.0'   '100.000'
alert(110..toFixed(1))      //     '110'   '110.0'
alert(110..toFixed(2))      //     '110'   '110.00'
alert(110..toFixed(3))      //   '110.0'   '110.000'
alert(110..toFixed(4))      //  '110.00'   '110.000'
alert(1110..toFixed(4))     //  '1110.0'   '1110.0000'
alert(11110..toFixed(4))    //   '11110'   '11110.0000'
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false
    },
    "target": "es2022",
    "loose": false,
    "minify": {
      "compress": {
        "arguments": false,
        "arrows": true,
        "booleans": true,
        "booleans_as_integers": false,
        "collapse_vars": true,
        "comparisons": true,
        "computed_props": true,
        "conditionals": true,
        "dead_code": true,
        "directives": true,
        "drop_console": false,
        "drop_debugger": true,
        "evaluate": true,
        "expression": false,
        "hoist_funs": false,
        "hoist_props": true,
        "hoist_vars": false,
        "if_return": true,
        "join_vars": true,
        "keep_classnames": false,
        "keep_fargs": true,
        "keep_fnames": false,
        "keep_infinity": false,
        "loops": true,
        "negate_iife": true,
        "properties": true,
        "reduce_funcs": false,
        "reduce_vars": false,
        "side_effects": true,
        "switches": true,
        "typeofs": true,
        "unsafe": false,
        "unsafe_arrows": false,
        "unsafe_comps": false,
        "unsafe_Function": false,
        "unsafe_math": false,
        "unsafe_symbols": false,
        "unsafe_methods": false,
        "unsafe_proto": false,
        "unsafe_regexp": false,
        "unsafe_undefined": false,
        "unused": true,
        "const_to_let": true,
        "pristine_globals": true
      },
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link

https://play.swc.rs/?version=1.3.35&code=H4sIAAAAAAAAA33RQQ6DIBAF0H1PwQ7dKAM9Q%2B9hlIWJSZuGJi56%2BLbCwFC%2FukAW84bPoNTx1%2Fe%2FdRjDa1jUW%2Fn14cfgp8uw%2BGdoqOvC%2FTavfmqobWujNHVG8z8BU4DZgW9pBKh837%2BUw%2F72FCDhzkUmBGMx0ZRuLWJZmCsKbU2ak91yqWQcTLYZ7WIl7woig7LlBzEcTWSriD0l2LjKpObFFATnlg%2Fiscm5VWSfTRJs%2FrNRzkbH2a4ScXeIkEpInBSVZMBxwHKp7D7nMRvXoAMAAA%3D%3D&config=H4sIAAAAAAAAA22TQZLjIAxF78K6F11ZzKIP0Ls%2BA0VAOGQwciGRjiuVu49wbIiT2VnPX0J8iZs6k1VfNzWZTJDrF82JzVV9KZ4nIJvDxOpDMQnyJhLcJTB5ABYJ0OHzcJDfEZFgFXyoMaTg51rM4jhlIKrfklRGSExNZ3LGXwk5F4mOiBFMeou1IR0SwwC5p1qM0UwE%2BmJyy6inmRwI0w4VBqenjNMTTS5wwCTFNubAOG3RQQMhg%2BVwgS6RGiJJJI21Thbo4FiGoRr4UMLFxGK41YLrYoOc2PJOGIi1L4le0K7TB3pcclUFrzNwyWnTnDGknQ9%2FAaTPaIiSGaFnLtzLHPZK%2Fx9VSF6GyHOjMuHeVYJBLqdD8O2GtWvIHLpbGVyxUG9oe%2FEV7i5EwYEG78Xtlky%2Fge2pF6vLiL6F4prxfQaPUG%2F7tKd1Bd7gt3TFz%2BNY%2BWj49MpoHo8Y30qMwCd0b1iMYHyFWbb3Or3SkhyIzeCefhSqYVtTGT6jjvW1bT7LQkiOHiIe2%2FrKoxxNGtpa3itAVyq4LeYtr%2FWPuvfXua0T%2FazCpdA%2FOW4Z9xIEAAA%3D

### Expected behavior

`0..toFixed(1)` should produce `0.0`

### Actual behavior

`0..toFixed(1)` produces `0`

### Version

1.3.35

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by magic-akari**
**Description:**

The code snippet  from StackOverflow is for handling `toPrecision` instead of `toFixed`

**BREAKING CHANGE:**

**Related issue (if exists):**
- fix #6957

### CODE REVIEW

**Comment by magic-akari** [comment]
The `num_to_precision` cannot handle JS `toPrecision`. It will return unexpected results.

```js
assert.sameValue((10).toPrecision(1), "1e+1");
assert.sameValue((11).toPrecision(1), "1e+1");
assert.sameValue((17).toPrecision(1), "2e+1");
assert.sameValue((19).toPrecision(1), "2e+1");
assert.sameValue((20).toPrecision(1), "2e+1");
```

after minify:
```js
assert.sameValue("10", "1e+1");
assert.sameValue("11", "1e+1");
assert.sameValue("17", "2e+1");
assert.sameValue("19", "2e+1");
assert.sameValue("20", "2e+1");
```

**Comment by kdy1** [comment]
Do you mean it's not possible at all?

**Comment by magic-akari** [comment]
> Do you mean it's not possible at all?

No, we can calculate the result step by step according to the ECMA262 specification.

**Comment by magic-akari** [comment]
There are two test case files from test262.
I don't know if there is a license issue. Is it OK to include them?

**Comment by kdy1** [comment]
I think it's fine

File: crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+assertEquals("0.000001ad7f29abcaf48", 0.0000001.toString(16));
+assertEquals("0.000002036565348d256", 0.00000012.toString(16));
+assertEquals("0.0000021047ee22aa466", 0.000000123.toString(16));
+assertEquals("0.0000002af31dc4611874", 0.00000001.toString(16));
+assertEquals("0.000000338a23b87483be", 0.000000012.toString(16));
+assertEquals("0.00000034d3fe36aaa0a2", 0.0000000123.toString(16));
+assertEquals("0", "0");
+assertEquals("-9", "-9");
+assertEquals("-5a", "-5a");
+assertEquals("-5a.1eb851eb852", -90.12.toString(16));
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: This seems wrong.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: I did
```js
function assertEquals(a, b) {
    if (a !== b) {
        throw `${a} !== ${b}`
    }    
}
```

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js (hunk 1 above)
Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: Weird, Chrome tells me the result is NaN.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: I found the problem.

```js
-90.12.toString(16)
```

is different from 
```js
(-90.12).toString(16)
```

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: Seems like `-` is processed after evaluating `.toString(16)`

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: Can you fix the fixer, too?

Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: > Can you fix the fixer, too?

Sure

Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: @kdy1 The root cause :
https://github.com/swc-project/swc/blob/ac9687af5f2e66499e7df9c0ac2f547e4494bd29/crates/swc_ecma_transforms_optimization/src/simplify/expr/mod.rs#L836-L843

optimize_unary_expr will convert 
`UnaryExpr(-, Lit::Num(90.12))` into `Lit::Num(-90.12)`.

Where should we fix it? Disable this behavior or fix it in `fixer`?

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:64]: I think disabling this behavior would be better, as it's sort of `invalid` state

File: crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+assertEquals("5yc1z", "5yc1z");
+assertEquals("0", "0");
+assertEquals("0", "0");
+assertEquals("0", "0");
+assertEquals("0", "0");
+assertEquals("0", "0");
+assertEquals("100000000000000000000000000000000", "100000000000000000000000000000000");
+assertEquals("100000000000000000000000000000001", "100000000000000000000000000000001");
+assertEquals("100000000000080", "100000000000080");
+assertEquals("1000000000000100", "1");
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:91]: I think this will fail

Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:91]:
```JavaScript 
assertEquals("1000000000000100", (-(-'0x1000000000000081')).toString(16));
```
Yes. It seems that somewhere the calculation result of `(-(-'0x1000000000000081'))` is wrong. 

[Check it](https://play.swc.rs/?version=1.3.35&code=H4sIAAAAAAAAA0ssLk4tKnEtLE3MKdZQMjRAAkCOko6Chq6GrrpBBYqMgYWhuqamXkl%2BcElRZl66hqGZpqY1AAUpUdlKAAAA&config=H4sIAAAAAAAAA22TQZLjIAxF78K6F11ZzKIP0Ls%2BA0VAOGQwciGRjiuVu49wbIiT2VnPX0J8iZs6k1VfNzWZTJDrF82JzVV9KZ4nIJvDxOpDMQnyJhLcJTB5ABYJ0OHzcJDfEZFgFXyoMaTg51rM4jhlIKrfklRGSExNZ3LGXwk5F4mOiBFMeou1IR0SwwC5p1qM0UwE%2BmJyy6inmRwI0w4VBqenjNMTTS5wwCTFNubAOG3RQQMhg%2BVwgS6RGiJJJI21Thbo4FiGoRr4UMLFxGK41YLrYoOc2PJOGIi1L4le0K7TB3pcclUFrzNwyWnTnDGknQ9%2FAaTPaIiSGaFnLtzLHPZK%2Fx9VSF6GyHOjMuHeVYJBLqdD8O2GtWvIHLpbGVyxUG9oe%2FEV7i5EwYEG78Xtlky%2Fge2pF6vLiL6F4prxfQaPUG%2F7tKd1Bd7gt3TFz%2BNY%2BWj49MpoHo8Y30qMwCd0b1iMYHyFWbb3Or3SkhyIzeCefhSqYVtTGT6jjvW1bT7LQkiOHiIe2%2FrKoxxNGtpa3itAVyq4LeYtr%2FWPuvfXua0T%2FazCpdA%2FOW4Z9xIEAAA%3D)

I can't locate the wrong code right now.
Maybe it should be fixed in another Pull Request.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/fixture/issues/6957/number-tostring/output.js:91]: I see, then let's postpone it

**Review by kdy1 — APPROVED** [review]
Thank you!


---

swc-bump:
 - swc_ecma_transforms_base

**Pull request merged**